### PR TITLE
Creating Revocation Flow for JWT One Time Token feature

### PR DESCRIPTION
**Pull request by GihanAyesh**
### Purpose
Fixes https://github.com/wso2/api-manager/issues/254

The [Design Documentation](https://docs.google.com/document/d/1gfhI9ozfHf8qV4SWRKfeUWFW8kSs0zAV2XFPnXGGiI0/edit?usp=sharing)

Please do not merge without the approval of @tharindu1st @HiranyaKavishani

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediator.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import org.wso2.carbon.apimgt.impl.factory.KeyManagerHolder;
+
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Service Implementation for One Time Token Revocation Mediator
+ */
+public class OneTimeTokenRevocationMediator extends AbstractMediator {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediator.java:39]: Override the **isContentAware** method and return false to not to build payload
   public boolean isContentAware() {
        return false;
    }

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/GatewayUtils.java [hunk 2]
```diff
@@ -1519,4 +1523,28 @@ public static List<String> getKeyManagers(org.apache.synapse.MessageContext mess
         }
         return Arrays.asList(APIConstants.KeyManager.API_LEVEL_ALL_KEY_MANAGERS);
     }
+
+    public static SignedJWTInfo getSignedJwt(String accessToken) throws ParseException {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/GatewayUtils.java:1527]: we can revert this change due to no usage of having public

**Review by tharindu1st — CHANGES_REQUESTED** [review]

**Review by tharindu1st — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    @Test
+    public void testMediateWithNullTokenScopes() {
+
+        List<String> nullScopes = new ArrayList<>();
+        Mockito.when(authContext.getRequestTokenScopes()).thenReturn(nullScopes);
+        Assert.assertTrue(mediator.mediate(messageContext));
+        Mockito.verify(authContext,Mockito.times(0)).
+                getConsumerKey();
+    }
+}
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java:154]: Please add a new line for EOF

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/kmclient/model/RevokeClient.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Key Manager endpoint to revoke One Time Tokens
+ */
+public interface RevokeClient {
+
+    @RequestLine("POST ")
+    @Headers("Content-Type: application/json")
+    Response revokeToken(@ApiParam("token") RevokeTokenInfoDTO token)
+            throws KeyManagerClientException;
+}
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/kmclient/model/RevokeClient.java:37]: please add a new line for EOF

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+     * JWT token does not have any scopes
+     */
+    @Test
+    public void testMediateWithNullTokenScopes() {
+
+        List<String> nullScopes = new ArrayList<>();
+        Mockito.when(authContext.getRequestTokenScopes()).thenReturn(nullScopes);
+        Assert.assertTrue(mediator.mediate(messageContext));
+        Mockito.verify(authContext,Mockito.times(0)).
+                getConsumerKey();
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java:152]: Move the code line to L151

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+    /**
+     * Issuer is set to null
+     */
+    @Test
+    public void testMediateWithNullIssuer() {
+
+        Mockito.when(authContext.getIssuer()).thenReturn(null);
+        Assert.assertTrue(mediator.mediate(messageContext));
+        Mockito.verify(authContext,Mockito.times(0)).
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java:113]: Move the code line to L151 as it doesn't exceed the max length

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+     * KeyManagerDto is set to null
+     */
+    @Test
+    public void testMediateWithNullKeyManagerDto() {
+
+        Mockito.when(KeyManagerHolder.getKeyManagerByIssuer(Mockito.anyString(), Mockito.anyString())).
+                thenReturn(null);
+        Assert.assertTrue(mediator.mediate(messageContext));
+        Mockito.verify(authContext,Mockito.times(0)).
+                getConsumerKey();
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java:127]: Move the code line to L151 as it doesn't exceed the max length

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+    /**
+     * AuthContext is set to null
+     */
+    @Test
+    public void testMediateWithNullAuthContext() {
+
+        Mockito.when(messageContext.getProperty(APISecurityUtils.API_AUTH_CONTEXT)).thenReturn(null);
+        Assert.assertTrue(mediator.mediate(messageContext));
+        Mockito.verify(authContext,Mockito.times(0)).
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediatorTest.java:101]: Move the code line to L151 as it doesn't exceed the max length

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediator.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+     */
+    public boolean mediate(MessageContext messageContext) {
+
+        AuthenticationContext authContext = (AuthenticationContext)
+                messageContext.getProperty(APISecurityUtils.API_AUTH_CONTEXT);
+        String tenantDomain = GatewayUtils.getTenantDomain();
+        if (authContext != null) {
+            String issuer = authContext.getIssuer();
+            List<String> scopes = authContext.getRequestTokenScopes();
+            if (StringUtils.isNotEmpty(issuer)) {
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/OneTimeTokenRevocationMediator.java:66]: What happens issuer is an empty string with spaces

**Pull request merged**